### ISSUE TITLE
Protect strong name key changes in ApiCompat / package validation

### ISSUE THREAD

**Issue by ViktorHofer** [dotnet/runtime#45899]
For System.Composition* packages we accidentally changed the strong name key in 5.0 and just reverted that: https://github.com/dotnet/runtime/pull/45756. We should add validation to prevent such mistakes in the future.

**Comment by ericstj** [comment]
Confirmed that APICompat doesn't catch this:
```
C:\Users\erics\.nuget\packages>dotnet microsoft.dotnet.apicompat\6.0.0-beta.20601.2\tools\netcoreapp2.1\Microsoft.DotNet.ApiCompat.dll system.composition.attributedmodel\1.4.1\lib\netstandard2.0\System.Composition.AttributedModel.dll -i system.composition.attributedmodel\5.0.0\lib\netstandard2.0\ --warn-on-missing-assemblies
Total Issues: 0

C:\Users\erics\.nuget\packages>sn -T system.composition.attributedmodel\1.4.1\lib\netstandard2.0\System.Composition.AttributedModel.dll

Microsoft (R) .NET Framework Strong Name Utility  Version 4.0.30319.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Public key token is b03f5f7f11d50a3a

C:\Users\erics\.nuget\packages>sn -T system.composition.attributedmodel\5.0.0\lib\netstandard2.0\System.Composition.AttributedModel.dll

Microsoft (R) .NET Framework Strong Name Utility  Version 4.0.30319.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Public key token is cc7b13ffcd2ddd51
```

**Comment by safern** [comment]
@ericstj should we just make this part of the roslyn based genAPI? instead of fixing it on the current one? If so, I'll self assign and add it to my GenAPI notes.

**Comment by ViktorHofer** [comment]
If we add this validation in existing tooling (GenAPI) then we could enable it in release/5.0 as well. That said, it's highly unlikely that we would unintentionally change a public key token in a servicing release without anyone noticing.

**Comment by safern** [comment]
> If we add this validation in existing tooling (GenAPI) then we could enable it in release/5.0 as well.

I think the break was due to a major refactor (removing DIrectory.Build.props) which shouldn't be accepted on release/5.0. My question was more like, is it worth it implementing it on both sides, new GenAPI and old GenAPI? 

I guess comparing public key tokens at assembly level should be "simple" to implement on existing GenAPI though.

**Comment by ViktorHofer** [comment]
> I think the break was due to a major refactor (removing DIrectory.Build.props) which shouldn't be accepted on release/5.0.

The regression happened in https://github.com/dotnet/corefx/commit/9a4ed0d00bc642b015974684c32b00650426429f#diff-739e1ffbaee417f430131875c0471f5ab63a11561f06de857adcd32f4eddce15L5 which didn't remove any Directory.Build.props file but consolidated the AssemblyVersion properties. The problem was that unintentionally the line that specified the public key token was removed as well. Servicing changes usually touch `PackageVersion` or `AssemblyVersion` properties for individual libraries (i.e. the current one for 5.0.2).

**Comment by safern** [comment]
> The problem was that unintentionally the line that specified the public key token was removed as well. Servicing changes usually touch PackageVersion or AssemblyVersion properties for individual libraries (i.e. the current one for 5.0.2).

Right, what I meant it was a major refactor to share `AssemblyVersion` properties.

**Comment by ViktorHofer** [comment]
Setting milestone to 6.0.0 as we believe we want to tackle this as part of the new ApiCompat work.
